src/Language/Futhark/Interpreter/AD.hs

```diff
@@ -0,0 +1,289 @@
+{-# LANGUAGE LambdaCase #-}
+module Language.Futhark.Interpreter.AD
+  (
+    ADSeed (..),
+    ADValue (..),
+    Op (..),
+    Tape (..),
+
+    value,
+    depth,
+    primal,
+
+    deriveVjp,
+
+    valueAsType,
+
+    addFor,
+
+    doOp,
+    handleOp,
+  )
+where
+
+import Language.Futhark.Primitive
+import Data.List
+import Data.Map qualified as M
+import Data.Set qualified as S
+
+import Futhark.AD.Derivatives
+import Futhark.Analysis.PrimExp
+import Language.Futhark.Core
+
+import Debug.Trace qualified as DBG
+import Control.Monad (zipWithM)
+import Data.Maybe (fromJust)
+import Data.Foldable (foldlM)
+
+data ADSeed
+  = VjpSeed Int Tape
+  | JvpSeed Int ADValue ADValue
+
+instance Show ADSeed where
+  show :: ADSeed -> String
+  show (VjpSeed d v) = "VjpSeed d" <> show d <> " " <> show v
+  show (JvpSeed d v dv) = "JvpSeed d" <> show d <> " " <> show v <> " " <> show dv
+
+data ADValue
+  = Primal PrimValue
+  | Seed ADSeed
+
+instance Show ADValue where
+  show (Primal v) = "Primal " <> show v
+  show (Seed s) = "Seed " <> show s
+
+data Op
+  = OpBin  BinOp
+  | OpCmp  CmpOp
+  | OpUn   UnOp
+  | OpFn   String
+  | OpConv ConvOp
+
+instance Show Op where
+  show (OpBin  op) = show op
+  show (OpCmp  op) = show op
+  show (OpUn   op) = show op
+  show (OpFn   fn) = show fn
+  show (OpConv op) = show op
+
+value :: ADValue -> PrimValue
+value (Primal v) = v
+value (Seed s) = value $ primal s
+
+depth :: ADSeed -> Int
+depth (VjpSeed d _) = d
+depth (JvpSeed d _ _) = d
+
+primal :: ADSeed -> ADValue
+primal (VjpSeed _ v) = tapeValue v
+primal (JvpSeed _ v _) = v
+
+valueAsType :: PrimValue -> Rational -> PrimValue
```

Comment:
You only ever call this function with a zero value for the `Rational`. There is a notion of `blankPrimValue` that you could use instead.